lane.py

```diff
@@ -991,105 +995,83 @@ def _groups_for_lanes(self, _db, relevant_lanes, queryable_lanes):
 
         queryable_lane_set = set(queryable_lanes)
 
-        qu = self._groups_query(_db, queryable_lanes)
-        if qu is None:
-            # We won't be running the query, but we still need to
-            # run this code because we may be making calls to groups()
-            # for WorkLists not handled by the query.
-            qu = []
-
-        # `items` is ordered with highest-quality items at the front.
-        # Go down the list of results, filling in the featured items
-        # for each lane until we have `target_size` items for each
-        # lane.
-        used_by_tier = defaultdict(list)
-        unused_by_tier = defaultdict(list)
-        used = set()
-        by_lane_id = defaultdict(list)
-        total_size = 0
-        total_parent_lane_size = 0
-
-        # Establish the maximum number of works we might need to pull
-        # from this query's results. If we ever have this many, we're
-        # done with that part of the work.
-        maximum_size = target_size * len(queryable_lane_set)
-        used_by_lane = defaultdict(set)
-        for mw, quality_tier, lane_id in qu:
-            if len(by_lane_id[lane_id]) >= target_size:
-                # We already have enough works to fill the lane for
-                # which this work was selected.
-                if parent_lane and lane_id != parent_lane.id:
-                    if mw.works_id not in used:
-                        # But this MaterializedWorkWithGenre hasn't
-                        # already been used in some other lane. Add
-                        # this to the 'unused' dictionary in case we
-                        # need to use it to fill in the parent lane.
-                        #
-                        # NOTE: Checking `used`, as we did above,
-                        # isn't totally reliable because this work
-                        # might show up again in another lane and get
-                        # used then, but it doesn't hurt.
-                        unused_by_tier[quality_tier].append(mw)
-                    continue
+        work_quality_tier_lane = list(
```

Comment:
This is new code that achieves the same basic result. It's simpler because there's no attempt to fill up the last lane with leftovers.